pkg/southbound/synchronizer/session.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2020-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package synchronizer
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+
+	"github.com/onosproject/onos-config/pkg/utils"
+
+	devicechange "github.com/onosproject/onos-config/api/types/change/device"
+	"github.com/onosproject/onos-config/pkg/dispatcher"
+	"github.com/onosproject/onos-config/pkg/events"
+	"github.com/onosproject/onos-config/pkg/modelregistry"
+	"github.com/onosproject/onos-config/pkg/southbound"
+
+	devicestore "github.com/onosproject/onos-config/pkg/store/device"
+	"github.com/onosproject/onos-config/pkg/store/mastership"
+
+	devicetype "github.com/onosproject/onos-config/api/types/device"
+	"github.com/onosproject/onos-config/pkg/store/change/device"
+	topodevice "github.com/onosproject/onos-topo/api/device"
+)
+
+const (
+	backoffInterval   = 10 * time.Millisecond
+	maxBackoffTime    = 5 * time.Second
+	mastershipTermKey = "onos-config.mastership.term"
+)
+
+// Session a gNMI session
+type Session struct {
+	mastershipStore           mastership.Store
+	deviceStore               devicestore.Store
+	mastershipState           *mastership.Mastership
+	connected                 bool
+	opStateChan               chan<- events.OperationalStateEvent
+	deviceResponseChan        chan events.DeviceResponse
+	dispatcher                *dispatcher.Dispatcher
+	modelRegistry             *modelregistry.ModelRegistry
+	operationalStateCache     map[topodevice.ID]devicechange.TypedValueMap
+	operationalStateCacheLock *sync.RWMutex
+	deviceChangeStore         device.Store
+	device                    *topodevice.Device
+	target                    southbound.TargetIf
+	cancel                    context.CancelFunc
+	closed                    bool
+	mu                        sync.RWMutex
+}
+
+func (s *Session) getCurrentTerm() (int, error) {
+	device, err := s.deviceStore.Get(s.device.ID)
+	if err != nil {
+		return 0, err
+	}
+
+	term := device.Attributes[mastershipTermKey]
+	if term == "" {
+		return 0, nil
+	}
+	return strconv.Atoi(term)
+}
+
+// open open a new gNMI session
+func (s *Session) open() error {
+	s.deviceResponseChan = make(chan events.DeviceResponse)
+
+	go func() {
+		_ = s.updateDeviceState()
+
+	}()
+	go func() {
+		s.mu.Lock()
+		s.connected = false
+		s.mu.Unlock()
+		state, _ := s.mastershipStore.GetMastership(s.device.ID)
```

Comment:
The session should be created with a mastership state. It shouldn't need to read the mastership state at all. That creates another race where the mastership could change between the time the session was created and the time it was opened. Just remove the `mastershipStore` field.

Reply:
@kuujo 
Done. I hope I addressed it properly.